database/user.go

```diff
@@ -166,3 +229,265 @@ func (db *DB) managedUsersByField(ctx context.Context, fieldName, fieldValue str
+			{"foreignField", "_id"},      // field in the skylinks collection
+			{"as", "fromSkylinks"},
+		}},
+	}
+	replaceStage := bson.D{
+		{"$replaceRoot", bson.D{
+			{"newRoot", bson.D{
+				{"$mergeObjects", bson.A{
+					bson.D{{"$arrayElemAt", bson.A{"$fromSkylinks", 0}}}, "$$ROOT"},
+				},
+			}},
+		}},
+	}
+	// This stage checks if the download has a non-zero `bytes` field and if so,
+	// it takes it as the download's size. Otherwise it reports the full
+	// skylink's size as download's size.
+	projectStage := bson.D{{"$project", bson.D{
+		{"size", bson.D{
+			{"$cond", bson.A{
+				bson.D{{"$gt", bson.A{"$bytes", 0}}}, // if
+				"$bytes",                             // then
+				"$size",                              // else
+			}},
+		}},
+	}}}
+
+	pipeline := mongo.Pipeline{matchStage, lookupStage, replaceStage, projectStage}
+	c, err := db.staticDownloads.Aggregate(ctx, pipeline)
+	if err != nil {
+		return 0, errors.AddContext(err, "DB query failed")
+	}
+	defer func() { _ = c.Close(ctx) }()
+	var bandwidth int64
+	// We need this struct, so we can safely decode both int32 and int64.
+	result := struct {
+		Size int64 `bson:"size"`
+	}{}
+	for c.Next(ctx) {
+		if err = c.Decode(&result); err != nil {
+			return 0, errors.AddContext(err, "failed to decode DB data")
+		}
+		bandwidth += BandwidthDownloadCost(result.Size)
+	}
+	return bandwidth, nil
+}
+
+// userRegistryWriteBandwidth reports the bandwidth used by the user's registry
+// writes.
+func (db *DB) userRegistryWriteBandwidth(ctx context.Context, userId primitive.ObjectID, monthStart time.Time) (int64, error) {
+	matchStage := bson.D{{"$match", bson.D{
+		{"user_id", userId},
+		{"timestamp", bson.D{{"$gt", monthStart}}},
+	}}}
+	writes, err := count(ctx, db.staticRegistryWrites, matchStage)
+	if err != nil {
+		return 0, errors.AddContext(err, "failed to fetch registry write bandwidth")
+	}
+	return writes * PriceBandwidthRegistryWrite, nil
+}
+
+// userRegistryReadBandwidth reports the bandwidth used by the user's registry
+// reads.
+func (db *DB) userRegistryReadBandwidth(ctx context.Context, userId primitive.ObjectID, monthStart time.Time) (int64, error) {
+	matchStage := bson.D{{"$match", bson.D{
+		{"user_id", userId},
+		{"timestamp", bson.D{{"$gt", monthStart}}},
+	}}}
+	reads, err := count(ctx, db.staticRegistryReads, matchStage)
+	if err != nil {
+		return 0, errors.AddContext(err, "failed to fetch registry read bandwidth")
+	}
+	return reads * PriceBandwidthRegistryRead, nil
+}
+
+// monthStart returns the start of the user's subscription month. Users get
+// their bandwidth quota reset at the start of the month.
+func (db *DB) monthStart(ctx context.Context, userId primitive.ObjectID) (time.Time, error) {
+	user, err := db.UserByID(ctx, userId)
+	if err != nil {
+		return time.Time{}, errors.AddContext(err, "failed to fetch user")
+	}
+	now := time.Now().UTC()
+	daysDelta := now.Day() - user.SubscribedUntil.Day()
```

Comment:
Time is hard. :D

Yes, there was a bug. I hope I managed to fix it. I also added a lengthy comment.

Here is how it looks now:
```go
func MonthStart(subscribedUntil time.Time) time.Time {
	now := time.Now().UTC()
	// Check how many days are left until the end of the user's subscription
	// month. Then calculate when the last subscription month started. We don't
	// care if the user is no longer subscribed and their sub expired 3 months
	// ago, all we care about here is the day of the month on which that
	// happened because that is the day from which we count their statistics for
	// the month. If they were never subscribed we use Jan 1st 1970 for
	// SubscribedUntil.
	daysDelta := subscribedUntil.Day() - now.Day()
	d := now.AddDate(0, -1, daysDelta)
	return time.Date(d.Year(), d.Month(), d.Day(), 0, 0, 0, 0, time.UTC)
}
```